// @react-native-community/cli/build/tools/packageManager.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setProjectDir = setProjectDir;
exports.install = install;
exports.installDev = installDev;
exports.uninstall = uninstall;
exports.installAll = installAll;

function _execa() {
  const data = _interopRequireDefault(require("execa"));

  _execa = function () {
    return data;
  };

  return data;
}

var _logger = _interopRequireDefault(require("./logger"));

var _yarn = require("./yarn");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let projectDir;

function executeCommand(command, args, options) {
  return (0, _execa().default)(command, args, {
    stdio: options && options.silent && !_logger.default.isVerbose() ? 'pipe' : 'inherit',
    cwd: options && options.cwd
  });
}

function shouldUseYarn(options) {
  if (options && options.preferYarn !== undefined) {
    return options.preferYarn && (0, _yarn.getYarnVersionIfAvailable)();
  }

  return (0, _yarn.isProjectUsingYarn)(projectDir) && (0, _yarn.getYarnVersionIfAvailable)();
}

function setProjectDir(dir) {
  projectDir = dir;
}

function install(packageNames, options) {
  return shouldUseYarn(options) ? executeCommand('yarn', ['add', ...packageNames], options) : executeCommand('npm', ['install', ...packageNames, '--save', '--save-exact'], options);
}

function installDev(packageNames, options) {
  return shouldUseYarn(options) ? executeCommand('yarn', ['add', '-D', ...packageNames], options) : executeCommand('npm', ['install', ...packageNames, '--save-dev', '--save-exact'], options);
}

function uninstall(packageNames, options) {
  return shouldUseYarn(options) ? executeCommand('yarn', ['remove', ...packageNames], options) : executeCommand('npm', ['uninstall', ...packageNames, '--save'], options);
}

function installAll(options) {
  return shouldUseYarn(options) ? executeCommand('yarn', ['install'], options) : executeCommand('npm', ['install'], options);
}
