// @react-native-community/cli/build/commands/init/editTemplate.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.changePlaceholderInTemplate = changePlaceholderInTemplate;

function _fsExtra() {
  const data = _interopRequireDefault(require("fs-extra"));

  _fsExtra = function () {
    return data;
  };

  return data;
}

function _path() {
  const data = _interopRequireDefault(require("path"));

  _path = function () {
    return data;
  };

  return data;
}

var _walk = _interopRequireDefault(require("../../tools/walk"));

function _cliTools() {
  const data = require("@react-native-community/cli-tools");

  _cliTools = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function replaceNameInUTF8File(filePath, projectName, templateName) {
  _cliTools().logger.debug(`Replacing in ${filePath}`);

  const content = _fsExtra().default.readFileSync(filePath, 'utf8').replace(new RegExp(templateName, 'g'), projectName).replace(new RegExp(templateName.toLowerCase(), 'g'), projectName.toLowerCase());

  _fsExtra().default.writeFileSync(filePath, content, 'utf8');
}

function renameFile(filePath, oldName, newName) {
  const newFileName = _path().default.join(_path().default.dirname(filePath), _path().default.basename(filePath).replace(new RegExp(oldName, 'g'), newName));

  _cliTools().logger.debug(`Renaming ${filePath} -> file:${newFileName}`);

  _fsExtra().default.moveSync(filePath, newFileName);
}

function shouldRenameFile(filePath, nameToReplace) {
  return _path().default.basename(filePath).includes(nameToReplace);
}

function shouldIgnoreFile(filePath) {
  return filePath.match(/node_modules|yarn.lock|package-lock.json/g);
}

const UNDERSCORED_DOTFILES = ['buckconfig', 'eslintrc.js', 'flowconfig', 'gitattributes', 'gitignore', 'watchmanconfig'];

function processDotfiles(filePath) {
  const dotfile = UNDERSCORED_DOTFILES.find(e => filePath.includes(`_${e}`));

  if (dotfile === undefined) {
    return;
  }

  renameFile(filePath, `_${dotfile}`, `.${dotfile}`);
}

function changePlaceholderInTemplate(projectName, placeholderName) {
  _cliTools().logger.debug(`Changing ${placeholderName} for ${projectName} in template`);

  (0, _walk.default)(process.cwd()).reverse().forEach(filePath => {
    if (shouldIgnoreFile(filePath)) {
      return;
    }

    if (!_fsExtra().default.statSync(filePath).isDirectory()) {
      replaceNameInUTF8File(filePath, projectName, placeholderName);
    }

    if (shouldRenameFile(filePath, placeholderName)) {
      renameFile(filePath, placeholderName, projectName);
    }

    if (shouldRenameFile(filePath, placeholderName.toLowerCase())) {
      renameFile(filePath, placeholderName.toLowerCase(), projectName.toLowerCase());
    }

    processDotfiles(filePath);
  });
}
